Java/core/src/main/java/com/amazon/randomcutforest/state/store/NodeStoreMapper.java

```diff
@@ -33,34 +37,132 @@
+        int leafFreeIndexPointer = state.getLeafFreeIndexPointer();
+
+        return new NodeStore(capacity, leftIndex, rightIndex, cutDimension, cutValue, leafMass, leafPointIndex,
+                nodeFreeIndexes, nodeFreeIndexPointer, leafFreeIndexes, leafFreeIndexPointer);
     }
 
     @Override
     public NodeStoreState toState(NodeStore model) {
         NodeStoreState state = new NodeStoreState();
-        state.setLeftIndex(Arrays.copyOf(model.leftIndex, model.leftIndex.length));
-        state.setRightIndex(Arrays.copyOf(model.rightIndex, model.rightIndex.length));
-        state.setParentIndex(Arrays.copyOf(model.parentIndex, model.parentIndex.length));
-        state.setMass(Arrays.copyOf(model.mass, model.mass.length));
-        state.setCutDimension(Arrays.copyOf(model.cutDimension, model.cutDimension.length));
-        state.setCutValue(Arrays.copyOf(model.cutValue, model.cutValue.length));
-        state.setFreeIndexes(Arrays.copyOf(model.getFreeIndexes(), model.getFreeIndexPointer() + 1));
+        state.setCapacity(model.getCapacity());
+        state.setCompressed(compress);
+        state.setSamplerNeeded(samplerNeeded);
+
+        BitSet bits = new BitSet(2 * model.getCapacity());
+        int[] leftIndex = Arrays.copyOf(model.leftIndex, model.leftIndex.length);
+        int[] rightIndex = Arrays.copyOf(model.rightIndex, model.rightIndex.length);
+        boolean check = compress && feasibleCanonical && reduceToBits(model.size(), leftIndex, rightIndex);
+        state.setCanonicalAndNotALeaf(check);
+        if (check) { // can have a canonical representation saving a lot of space
+            state.setLeftIndex(ArrayPacking.pack(leftIndex, state.isCompressed()));
+            state.setRightIndex(ArrayPacking.pack(rightIndex, state.isCompressed()));
+            state.setSize(model.size());
+        } else { // the temporary array leftIndex and rightIndex may be corrupt in reduceToBits()
+            state.setLeftIndex(ArrayPacking.pack(model.leftIndex, compress));
+            state.setRightIndex(ArrayPacking.pack(model.rightIndex, compress));
+        }
+
+        state.setCutDimension(ArrayPacking.pack(model.cutDimension, compress));
+        state.setCutValueDouble(Arrays.copyOf(model.cutValue, model.cutValue.length));
+
+        state.setNodeFreeIndexes(ArrayPacking.pack(model.getNodeFreeIndexes(), compress));
+        state.setNodeFreeIndexPointer(model.getNodeFreeIndexPointer());
+        state.setLeafFreeIndexes(ArrayPacking.pack(model.getLeafFreeIndexes(), compress));
+        state.setLeafFreeIndexPointer(model.getLeafFreeIndexPointer());
+        if (!samplerNeeded) {
+            state.setLeafPointIndex(ArrayPacking.pack(model.getLeafPointIndex(), compress));
+            state.setLeafmass(ArrayPacking.pack(model.getLeafMass(), compress));
+        }
+
         return state;
     }
+
+    boolean reduceToBits(int size, int[] leftIndex, int[] rightIndex) {
```

Comment:
It looks like this method is doing two things: encoding the left index and right index arrays to indicate whether a given node has children or not, and performing some correctness on the arrays. Currently I think certain kinds of bad inputs could still pass the checks in this method. I think we should either make this method fully validate the arrays or else we should remove the existing validation. The partial validation we're currently doing seems like it lead to trouble later.

My suggestion would be to add a `isCanonical()` or `isCanonicalized()` method to NodeStore and remove the validation from here. It will make creating the bit arrays simpler (although we'll end up doing two scans of the index arrays).

Reply:
Done. Maybe there can be a better reorganization of code down the road -- but perhaps not in this PR.